files/default/neutron-evacuate-lbaasv2-agent.py

```diff
@@ -0,0 +1,223 @@
+            "SELECT id,host,heartbeat_timestamp "
+            "FROM agents WHERE "
+            "agents.host != %s "
+            "AND agents.binary='neutron-lbaasv2-agent' "
+            "AND agents.admin_state_up=true;",
+            self.host_to_evacuate)
+
+        return [{'id': id, 'host': host}
+                for id, host, timestamp in ret.fetchall()
+                if not timeutils.is_older_than(timestamp,
+                                               cfg.CONF.agent_down_time)]
+
+    def loadbalancers_on_agent(self, agent_hostname):
+        ret = self.connection.execute(
+            "SELECT loadbalancer_id "
+            "FROM lbaas_loadbalanceragentbindings,agents "
+            "WHERE agents.host = %s "
+            "AND agents.binary='neutron-lbaasv2-agent' "
+            "AND agents.id=lbaas_loadbalanceragentbindings.agent_id;",
+            agent_hostname)
+        return [id[0] for id in ret.fetchall()]
+
+    def reassign_loadbalancers(self, load_balancer_ids, target_agents):
+        for agent in target_agents:
+            agent['load'] = len(self.loadbalancers_on_agent(agent['host']))
+
+        agents_to_restart = set()
+        for loadbalancer in load_balancer_ids:
+            agent = min(target_agents, key=lambda x: x['load'])
+            LOG.info("Reassigning loadbalancer %s to agent %s" %
+                     (loadbalancer, agent))
+
+            self.connection.execute(
+                "UPDATE lbaas_loadbalanceragentbindings "
+                "SET agent_id = %s "
+                "WHERE loadbalancer_id = %s;",
+                agent['id'], loadbalancer
+            )
+            agents_to_restart.add(agent['host'])
+            agent['load'] = agent['load'] + 1
+        return agents_to_restart
+
+    def setup(self):
+        cfg.CONF(project='neutron')
+        # Always log to stderr, ignore setting in neutron.conf
+        cfg.CONF.set_override("use_stderr", True)
+        logging.setup(cfg.CONF, "neutron")
+        self.host_to_evacuate = cfg.CONF.host
+
+        db_uri = cfg.CONF.database.connection
+        db = sqlalchemy.create_engine(db_uri)
+        try:
+            self.connection = db.connect()
+        except sqlalchemy.exc.SQLAlchemyError as e:
+            LOG.error('Cannot connect to database: %s' % e)
+            raise
+
+    def run(self):
+        load_balancers = self.loadbalancers_on_agent(self.host_to_evacuate)
+        if load_balancers:
+            LOG.info("Evacuating loadbalancers: %s", load_balancers)
+            target_agents = self.available_destination_agents()
+            LOG.info("Available destination agents %s", target_agents)
+            agents_to_restart = list(
+                self.reassign_loadbalancers(load_balancers, target_agents)
+            )
+
+            if cfg.CONF.source_agent_restart:
+                # Make sure the source agent is restarted first
+                agents_to_restart.insert(0, self.host_to_evacuate)
```

Comment:
Yes this is a upgrade specific feature. (The source agents are down after the first network node is upgraded)

But the default for "cfg.CONF.source_agent_restart" is acatually true. When needed we call the tool with "--nosource_agent_restart".